File: src/CSharp/801-900/821.ShortestToChar.cs

using System;
using System.Linq;
using System.Collections.Generic;

namespace PeefyLeetCode.ShortestToChar
{
public class Solution {
    public int[] ShortestToChar(string S, char C) {
        var indexlist = new List<int>();
        int[] r = new int[S.Length];   
        int j = 1;
        for (int i = 0; i < S.Length; i++)
        {
            if (C == S[i])
                indexlist.Add(i);
        }
        indexlist.Insert(0, -10000);
        indexlist.Add(10000); 
        for (int i = 0; i < S.Length; i++)
        {
            if (C == S[i]){
                ++j;
            }
            else
                r[i] = Math.Min(i - indexlist[j - 1], indexlist[j] - i);
        }
        return r;
    }
}
}

File: src/Python/801-900/821.ShortestToChar.py

class Solution:
    def shortestToChar(self, S, C):
        """
        :type S: str
        :type C: str
        :rtype: List[int]
        """
        indexlist = []
        r = []
        for i, s in enumerate(S):
            if C == s:
                indexlist.append(i)
        indexlist = [-10000] + indexlist + [10000]
        j = 1
        for i, s in enumerate(S):
            if C == s:
                r.append(0)
                j += 1
            else:           
                r.append(min(i - indexlist[j - 1], indexlist[j] - i))   
        return r
        
if __name__ == '__main__':
    solution = Solution()
    print(solution.shortestToChar(S = "loveleetcode", C = 'e'))
else:
    pass




